xdp/src/netlink.rs

```diff
@@ -133,10 +124,120 @@ impl NetlinkSocket {
 
         Ok(messages)
     }
+
+    /// Opens a listener socket for netlink updates
+    /// NETLINK_ROUTE socket subscribed to `groups` bitmask
+    pub fn bind(groups: u32) -> Result<Self, io::Error> {
+        let mut sock = Self::open()?;
+
+        // Subscribe to multicast groups
+        let mut addr: sockaddr_nl = unsafe { mem::zeroed() };
+        addr.nl_family = AF_NETLINK as u16;
+        addr.nl_groups = groups;
+        if unsafe {
+            libc::bind(
+                sock.as_raw_fd(),
+                &addr as *const _ as *const _,
+                mem::size_of::<sockaddr_nl>() as u32,
+            )
+        } < 0
+        {
+            return Err(io::Error::last_os_error());
+        }
+
+        // Larger rcvbuf to survive bursts of updates
+        let rcvbuf: i32 = 1 << 16;
+        unsafe {
+            setsockopt(
+                sock.as_raw_fd(),
+                SOL_SOCKET,
+                SO_RCVBUF,
+                &rcvbuf as *const _ as *const _,
+                mem::size_of::<i32>() as u32,
+            );
+        }
+
+        sock.nl_pid = Self::get_nl_pid(sock.as_raw_fd())?;
+
+        Ok(sock)
+    }
+
+    #[inline]
+    fn create_raw_socket() -> Result<OwnedFd, io::Error> {
+        let raw = unsafe { socket(AF_NETLINK, SOCK_RAW, NETLINK_ROUTE) };
+        if raw < 0 {
+            return Err(io::Error::last_os_error());
+        }
+        let sock = unsafe { OwnedFd::from_raw_fd(raw) };
+
+        let enable = 1i32;
+        if unsafe {
+            setsockopt(
+                sock.as_raw_fd(),
+                SOL_NETLINK,
+                NETLINK_EXT_ACK,
+                &enable as *const _ as *const _,
+                mem::size_of::<i32>() as u32,
+            )
+        } < 0
+        {
+            return Err(io::Error::last_os_error());
+        }
+        Ok(sock)
+    }
+
+    #[inline]
+    fn get_nl_pid(fd: std::os::fd::RawFd) -> Result<u32, io::Error> {
+        let mut name: sockaddr_nl = unsafe { mem::zeroed() };
+        let mut len = mem::size_of::<sockaddr_nl>() as u32;
+        if unsafe { getsockname(fd, &mut name as *mut _ as *mut _, &mut len as *mut _) } < 0 {
+            return Err(io::Error::last_os_error());
+        }
+        Ok(name.nl_pid)
+    }
+
+    #[inline]
+    pub fn as_raw_fd(&self) -> std::os::fd::RawFd {
+        self.sock.as_raw_fd()
+    }
+}
+
+#[inline]
+pub(crate) fn is_supported_ipv4_route_header(msg: &NetlinkMessage) -> bool {
```

Comment:
looks like this should be in route_monitor.rs?

Reply:
also, there is duplication in the checks between this function and
is_valid_route(), can we do something smarter?